File: cumulative-core/src/main/java/io/github/reggert/cumulative/core/scan/ScannerSettings.java
package io.github.reggert.cumulative.core.scan;

import org.apache.accumulo.core.security.Authorizations;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.Serializable;
import java.time.Duration;
import java.util.*;

import static java.util.Objects.requireNonNull;


/**
 * Immutable representation of settings to use for a scan.
 */
public abstract class ScannerSettings implements Serializable {
    private static final long serialVersionUID = 1L;
    private final Duration timeout;
    private final Duration batchTimeout;
    private final String classloaderContext;
    private final Authorizations authorizations;


    public static final class Simple extends ScannerSettings {
        private static final long serialVersionUID = 1L;
        private final Integer batchSize;
        private final Long readAheadThreshold;
        private final boolean isolationEnabled;
        public static final Simple DEFAULT = builder().build();

        private Simple(
            @Nullable final Duration timeout,
            @Nullable final Duration batchTimeout,
            @Nullable final String classloaderContext,
            @Nonnull final Authorizations authorizations,
            @Nullable final Integer batchSize,
            @Nullable final Long readAheadThreshold,
            final boolean isolationEnabled
        ) {
            super(timeout, batchTimeout, classloaderContext, authorizations);
            this.batchSize = batchSize;
            this.readAheadThreshold = readAheadThreshold;
            this.isolationEnabled = isolationEnabled;
        }

        public Optional<Integer> getBatchSize() {
            return Optional.ofNullable(batchSize);
        }

        public Optional<Long> getReadAheadThreshold() {
            return Optional.ofNullable(readAheadThreshold);
        }

        public boolean isIsolationEnabled() {
            return isolationEnabled;
        }

        @Override
        public boolean equals(final Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            if (!super.equals(o)) return false;
            final Simple that = (Simple) o;
            return this.isolationEnabled == that.isolationEnabled &&
                Objects.equals(this.batchSize, that.batchSize) &&
                Objects.equals(this.readAheadThreshold, that.readAheadThreshold);
        }

        @Override
        public int hashCode() {
            return Objects.hash(super.hashCode(), batchSize, readAheadThreshold, isolationEnabled);
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("ScannerSettings.Simple{");
            sb.append("timeout=").append(getTimeout().orElse(null));
            sb.append(", batchTimeout=").append(getBatchTimeout().orElse(null));
            sb.append(", classloaderContext='").append(getClassloaderContext().orElse(null)).append('\'');
            sb.append(", authorizations=").append(getAuthorizations());
            sb.append(", batchSize=").append(batchSize);
            sb.append(", readAheadThreshold=").append(readAheadThreshold);
            sb.append(", isolationEnabled=").append(isolationEnabled);
            sb.append('}');
            return sb.toString();
        }

        public static final class Builder extends ScannerSettings.Builder {
            private Integer batchSize;
            private Long readAheadThreshold;
            private boolean isolationEnabled;

            private Builder() {
                this.batchSize = null;
                this.readAheadThreshold = null;
                this.isolationEnabled = false;
            }

            private Builder(@Nonnull final ScannerSettings.Simple source) {
                super(source);
                this.batchSize = source.batchSize;
                this.readAheadThreshold = source.readAheadThreshold;
                this.isolationEnabled = source.isolationEnabled;
            }

            public Builder batchSize(final int batchSize) {
                this.batchSize = batchSize;
                return this;
            }

            public Builder clearBatchSize() {
                this.batchSize = null;
                return this;
            }

            public Builder readAheadThreshold(final long readAheadThreshold) {
                this.readAheadThreshold = readAheadThreshold;
                return this;
            }

            public Builder clearReadAheadThreshold() {
                this.readAheadThreshold = null;
                return this;
            }

            public Builder isolationEnabled(final boolean isolationEnabled) {
                this.isolationEnabled = isolationEnabled;
                return this;
            }

            public Simple build() {
                return new ScannerSettings.Simple(
                    timeout,
                    batchTimeout,
                    classloaderContext,
                    authorizations,
                    batchSize,
                    readAheadThreshold,
                    isolationEnabled
                );
            }
        }

        /**
         * Constructs a new builder with default settings.
         *
         * @return a new {@code builder}.
         */
        public static Builder builder() {
            return new Builder();
        }


        /**
         * Copies the settings from this object into a new builder.
         *
         * @return a new {@code Builder}.
         */
        public Builder toBuilder() {
            return new Builder(this);
        }
    }


    public static final class Batch extends ScannerSettings {
        private static final long serialVersionUID = 1L;
        private final int numberOfQueryThreads;
        public static final int DEFAULT_QUERY_THREADS = 2;
        public static final Batch DEFAULT = builder().build();

        private Batch(
            @Nullable final Duration timeout,
            @Nullable final Duration batchTimeout,
            @Nullable final String classloaderContext,
            @Nonnull final Authorizations authorizations,
            final int numberOfQueryThreads
        ) {
            super(timeout, batchTimeout, classloaderContext, authorizations);
            this.numberOfQueryThreads = numberOfQueryThreads;
        }

        public int getNumberOfQueryThreads() {
            return numberOfQueryThreads;
        }

        @Override
        public boolean equals(final Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            if (!super.equals(o)) return false;
            final Batch batch = (Batch) o;
            return numberOfQueryThreads == batch.numberOfQueryThreads;
        }

        @Override
        public int hashCode() {

            return Objects.hash(super.hashCode(), numberOfQueryThreads);
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("ScannerSettings.Batch{");
            sb.append("timeout=").append(getTimeout().orElse(null));
            sb.append(", batchTimeout=").append(getBatchTimeout().orElse(null));
            sb.append(", classloaderContext='").append(getClassloaderContext().orElse(null)).append('\'');
            sb.append(", authorizations=").append(getAuthorizations());
            sb.append(", numberOfQueryThreads=").append(numberOfQueryThreads);
            sb.append('}');
            return sb.toString();
        }

        public static final class Builder extends ScannerSettings.Builder {
            private int numberOfQueryThreads;

            private Builder() {
                this.numberOfQueryThreads = DEFAULT_QUERY_THREADS;
            }

            private Builder(@Nonnull final ScannerSettings.Batch source) {
                super(source);
                this.numberOfQueryThreads = source.numberOfQueryThreads;
            }

            public Builder numberOfQueryThreads(final int numberOfQueryThreads) {
                this.numberOfQueryThreads = numberOfQueryThreads;
                return this;
            }

            public Batch build() {
                return new ScannerSettings.Batch(
                    timeout,
                    batchTimeout,
                    classloaderContext,
                    authorizations,
                    numberOfQueryThreads
                );
            }
        }

        /**
         * Constructs a new builder with default settings.
         *
         * @return a new {@code builder}.
         */
        public static Builder builder() {
            return new Builder();
        }


        /**
         * Copies the settings from this object into a new builder.
         *
         * @return a new {@code Builder}.
         */
        public Builder toBuilder() {
            return new Builder(this);
        }
    }


    public static abstract class Builder {
        protected Duration timeout;
        protected Duration batchTimeout;
        protected String classloaderContext;
        protected Authorizations authorizations;

        private Builder() {
            this.timeout = null;
            this.batchTimeout = null;
            this.classloaderContext = null;
            this.authorizations = new Authorizations();
        }

        private Builder(@Nonnull final ScannerSettings source) {
            this.timeout = source.timeout;
            this.batchTimeout = source.batchTimeout;
            this.classloaderContext = source.classloaderContext;
            this.authorizations = source.authorizations;
        }

        public Builder timeout(@Nonnull final Duration timeout) {
            requireNonNull(timeout, "timeout");
            if (timeout.isNegative() || timeout.isZero()) {
                throw new IllegalArgumentException("timeout must be positive");
            }
            this.timeout = timeout;
            return this;
        }

        public Builder clearTimeout() {
            this.timeout = null;
            return this;
        }

        public Builder batchTimeout(@Nonnull final Duration batchTimeout) {
            requireNonNull(batchTimeout, "batchTimeout");
            if (timeout.isNegative() || timeout.isZero()) {
                throw new IllegalArgumentException("batchTimeout must be positive");
            }
            this.batchTimeout = batchTimeout;
            return this;
        }

        public Builder clearBatchTimeout() {
            this.batchTimeout = null;
            return this;
        }

        public Builder classloaderContext(@Nonnull final String classloaderContext) {
            this.classloaderContext = requireNonNull(classloaderContext, "classloaderContext");
            return this;
        }

        public Builder clearClassloaderContext() {
            this.classloaderContext = null;
            return this;
        }
    }


    private ScannerSettings(
        @Nullable final Duration timeout,
        @Nullable final Duration batchTimeout,
        @Nullable final String classloaderContext,
        @Nonnull final Authorizations authorizations
    ) {
        this.timeout = timeout;
        this.batchTimeout = batchTimeout;
        this.classloaderContext = classloaderContext;
        this.authorizations = authorizations;
    }


    public Optional<Duration> getTimeout() {
        return Optional.ofNullable(timeout);
    }


    public Optional<Duration> getBatchTimeout() {
        return Optional.ofNullable(batchTimeout);
    }


    public Optional<String> getClassloaderContext() {
        return Optional.ofNullable(classloaderContext);
    }


    public Authorizations getAuthorizations() {
        return authorizations;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final ScannerSettings that = (ScannerSettings) o;
        return Objects.equals(this.timeout, that.timeout) &&
            Objects.equals(this.batchTimeout, that.batchTimeout) &&
            Objects.equals(this.classloaderContext, that.classloaderContext) &&
            Objects.equals(this.authorizations, that.authorizations);
    }


    @Override
    public int hashCode() {
        return Objects.hash(timeout, batchTimeout, classloaderContext, authorizations);
    }

}

File: cumulative-core/src/main/scala/io/github/reggert/cumulative/core/scan/ScannerSettings.scala
package io.github.reggert.cumulative.core.scan

import java.util.concurrent.TimeUnit

import org.apache.accumulo.core.Constants
import org.apache.accumulo.core.client.mapreduce.{AbstractInputFormat, AccumuloInputFormat, InputFormatBase}
import org.apache.accumulo.core.client.{Scanner, ScannerBase}
import org.apache.accumulo.core.security.Authorizations
import org.apache.hadoop.mapreduce.Job

import scala.concurrent.duration.Duration


/**
  * Base class for containers of settings to pass to Accumulo scanners.
  */
sealed trait ScannerSettings extends Serializable {
  /**
    * The timeout for retries in the event of I/O failures; may be undefined/infinite to retry indefinitely.
    */
  val timeout : Duration

  /**
    * The amount of time to wait for the batch buffer to fill before returning the results; may be
    * undefined/infinite to always wait for the buffer to completely fill.
    */
  val batchTimeout : Duration

  /**
    * Optional classloader context to pass to the underlying scanner.
    */
  val classLoaderContext : Option[String]

  /**
    * Scan authorizations to pass to the scanner; this, intersected with the user's allowed authorizations,
    * is checked against each entry's visibility to determine whether to return that entry to the user.
    */
  val authorizations : Authorizations

  /**
    * Applies these settings to a scanner.
    *
    * @param scanner scanner to configure.
    */
  protected final def apply(scanner : ScannerBase) : Unit = {
    if (timeout.isFinite()) {
      scanner.setTimeout(timeout.toMillis, TimeUnit.MILLISECONDS)
    }
    if (batchTimeout.isFinite()) {
      scanner.setBatchTimeout(batchTimeout.toMillis, TimeUnit.MILLISECONDS)
    }
    classLoaderContext.foreach(scanner.setClassLoaderContext)
  }

  /**
    * Applies these settings to a Hadoop job configuration (for use with [[AccumuloInputFormat]]).
    *
    * @param configuration Hadoop job configuration to which to apply settings.
    */
  def apply(configuration : Job) : Unit = {
    classLoaderContext.foreach(AbstractInputFormat.setClassLoaderContext(configuration, _))
    AbstractInputFormat.setScanAuthorizations(configuration, authorizations)
  }
}


object ScannerSettings {

  /**
    * Settings used by simple scanners.
    *
    * @param timeout The timeout for retries in the event of I/O failures; may be undefined/infinite to retry
    *                indefinitely.
    * @param batchTimeout The amount of time to wait for the batch buffer to fill before returning the
    *                     results; may be undefined/infinite to always wait for the buffer to completely fill.
    * @param classLoaderContext Optional classloader context to pass to the underlying scanner.
    * @param authorizations Scan authorizations to pass to the scanner; this, intersected with the user's
    *                       allowed authorizations, s checked against each entry's visibility to determine
    *                       whether to return that entry to the user.
    * @param batchSize The number of entries to attempt to return from the server at a time.
    * @param readAheadThreshold The number of batches of entries the scanner will return before triggering
    *                           readahead in the background.
    * @param isolationEnabled Whether to enable scan isolation.
    */
  final case class Simple(
    timeout: Duration = Duration.Undefined,
    batchTimeout: Duration = Duration.Undefined,
    classLoaderContext: Option[String] = None,
    authorizations: Authorizations = new Authorizations,
    batchSize : Int = Constants.SCAN_BATCH_SIZE,
    readAheadThreshold : Long = Constants.SCANNER_DEFAULT_READAHEAD_THRESHOLD,
    isolationEnabled : Boolean = false
  ) extends ScannerSettings {
    require(batchSize > 0, s"batchSize ($batchSize) must be greater than 0")
    require(readAheadThreshold > 0L, s"readAheadThreshold ($readAheadThreshold) must be greater than 0")

    /**
      * Applies these settings to a [[Scanner]].
      *
      * @param scanner the scanner to configure.
      */
    def apply(scanner : Scanner) : Unit = {
      super.apply(scanner)
      scanner.setBatchSize(batchSize)
      scanner.setReadaheadThreshold(readAheadThreshold)
      if (isolationEnabled) {
        scanner.enableIsolation()
      }
    }

    override def apply(configuration : Job) : Unit = {
      super.apply(configuration)
      InputFormatBase.setScanIsolation(configuration, isolationEnabled)
    }
  }


  /**
    * Settings used by batch scanners.
    *
    * @param timeout The timeout for retries in the event of I/O failures; may be undefined/infinite to retry
    *                indefinitely.
    * @param batchTimeout The amount of time to wait for the batch buffer to fill before returning the
    *                     results; may be undefined/infinite to always wait for the buffer to completely fill.
    * @param classLoaderContext Optional classloader context to pass to the underlying scanner.
    * @param authorizations Scan authorizations to pass to the scanner; this, intersected with the user's
    *                       allowed authorizations, s checked against each entry's visibility to determine
    *                       whether to return that entry to the user.
    * @param numberOfQueryThreads Size of thread pool to use to query multiple tablet servers simultaneously.
    */
  final case class Batch(
    timeout: Duration = Duration.Undefined,
    batchTimeout: Duration = Duration.Undefined,
    classLoaderContext: Option[String] = None,
    authorizations: Authorizations = new Authorizations,
    numberOfQueryThreads : Int = 2
  ) extends ScannerSettings

}


